### PR TITLE
Implement caching types

### LINKED ISSUE #3: Cache-Control header handling
| Q | A |
| --- | --- |
| Bug? | no |
| New Feature? | yes |

See php-http/plugins#58

### PR DESCRIPTION
**Pull request by acrobat**
| Q               | A
| --------------- | ---
| Bug fix?        | yes and no
| New feature?    | yes
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | fixes #3 
| Documentation   | https://github.com/php-http/documentation/issues/178
| License         | MIT

This is a first simple version to get the discussion going and see what changes are needed here. 

#### What's in this PR?

As discussed in #3 this PR introduces 2 factory methods and an option to specify which cache control directives should be checked

#### Checklist

- [x] Updated CHANGELOG.md to describe BC| new feature / bugfix
- [x] Documentation pull request created https://github.com/php-http/documentation/issues/178
- [x] Add tests

### CODE REVIEW

File: src/CachePlugin.php
Comment (dbu) [inline src/CachePlugin.php]: we should check if the option already exists and add it to the existing list

**Review by dbu — COMMENTED** [review]
i find the name `excluded_directives` confusing. we seem to not exclude those directives, but use them.

i would call the option `respect_response_cache_directives` and deprecate `respect_cache_headers`.

File: src/CachePlugin.php
Comment (tuupola) [reply src/CachePlugin.php]: Or maybe just allow `respect_cache_headers` to be either boolean or array. If boolean true then assume it is all directives. If false then assume it is empty array.

Comment (dbu) [reply src/CachePlugin.php]: the things like no-cache are not headers, but directives. and some can be both on request and/or response, and we should treat that separately to allow the user full control. maybe they want no-cache on request to be respected, but on the response they still want to cache...

**Comment by acrobat** [comment]
@dbu thanks for the feedback! I've updated the code!

File: src/CachePlugin.php
Comment (acrobat) [reply src/CachePlugin.php]: I'm also more in favor of separate options. Because I would avoid mixing option types bool/array

**Comment by acrobat** [comment]
I've tested my patch on a project using the github api (which sends the private cache directive with each request). And I can confirm it works when you construct the cache plugin with the clientCache method.

The phpspec tests fail after the patch, but I will look into this fail later tonight!

File: src/CachePlugin.php
Comment (dbu) [inline src/CachePlugin.php]: i would do the cleanup (create an empty array in respect_response_cache_directives) of this in the constructor and issue a deprecation warning if respect_cache_headers is defined. if both are defined in the constructor, throw an exception as its wrong to do that.

File: src/CachePlugin.php
Comment (dbu) [inline src/CachePlugin.php]: this means that if max-age is set to be respected and defined, we do not cache. i think you need a constant for NO_CACHE_FLAGS and array intersect that with respect_response_cache_directives

File: src/CachePlugin.php
Comment (dbu) [inline src/CachePlugin.php]: lets add a phpdoc here that says this sets up the cache plugin in a client mode. it will cache even with cache-control private or no-store.

File: src/CachePlugin.php
Comment (dbu) [inline src/CachePlugin.php]: to stay BC, this should include no-store

**Review by dbu — CHANGES_REQUESTED** [review]
thanks, i think we move in the right direction with this. this will end up being quite flexible, while still being reasonably understandable.

File: src/CachePlugin.php
Comment (acrobat) [reply src/CachePlugin.php]: Yes you are right, I will look into a way to do this. We can't use a constant with an array value as the cacheplugin supports php >=5.4

Comment (dbu) [reply src/CachePlugin.php]: ah, right. then private property is fine too

**Comment by acrobat** [comment]
@dbu I've incorporated the feedback and I've added a phpspec test for this new option/behavior

File: src/CachePlugin.php
Comment (Nyholm) [inline src/CachePlugin.php]: "can be cached"? "noCacheFlags"?

You mean cannot be cached?

File: src/CachePlugin.php
Comment (Nyholm) [inline src/CachePlugin.php]: Generally. Make lines shorter than 100 chars. (PSR-2)

File: src/CachePlugin.php
Comment (Nyholm) [inline src/CachePlugin.php]: Same here.. Long lines.

File: src/CachePlugin.php
Comment (Nyholm) [inline src/CachePlugin.php]: "See config for constructor" or something similar.

File: src/CachePlugin.php
Comment (Nyholm) [inline src/CachePlugin.php]: Missing space

File: src/CachePlugin.php
Comment (Nyholm) [inline src/CachePlugin.php]: Correct me if Im wrong but one could simplify this by: 

```php
$config['respect_response_cache_directives'][] = 'no-cache';
$config['respect_response_cache_directives'][] = 'max-age';
array_unique($config['respect_response_cache_directives']);
```
Right?

File: src/CachePlugin.php
Comment (Nyholm) [inline src/CachePlugin.php]: Should you add an array here? `['no-cache']` vs `'no-cache'`.

File: src/CachePlugin.php
Comment (Nyholm) [inline src/CachePlugin.php]: Same here, document what parameters or refer to constructor

File: src/CachePlugin.php
Comment (Nyholm) [inline src/CachePlugin.php:186]: This cannot be removed. That will break BC.

File: src/CachePlugin.php
Comment (Nyholm) [inline src/CachePlugin.php:189]: Removing this will also break BC, right?

Can we do something smart in the constructor? If `respect_cache_headers` is true then populate `respect_response_cache_directives` with "no-store", "private".

File: src/CachePlugin.php
Comment (Nyholm) [inline src/CachePlugin.php]: $cacheableDirectives can only be one of these four (no matter what). Is this really the intention?

```php
 ['no-cache', 'private']
 ['private']
 ['no-cache']
[]
```

**Review by Nyholm — CHANGES_REQUESTED** [review]
Thank you for this PR. I really like the idea of these factory methods. 

I had some comments on some things I did not understand.

File: src/CachePlugin.php
Comment (acrobat) [reply src/CachePlugin.php]: The directive list contains other values like `max-age, no-store` and possibly other values in the future. So I've added this method to easily extract cache directives if they are provided in the config

File: src/CachePlugin.php
Comment (acrobat) [reply src/CachePlugin.php:189]: I don't think this breaks BC as the default options include the headers

File: src/CachePlugin.php
Comment (acrobat) [reply src/CachePlugin.php:186]: On line 361 I've added some BC logic. If this option is false we add an empty array so all the checks are skipped and original behavior is kept

**Comment by acrobat** [comment]
Thanks @Nyholm for the review! I've fixed your comments

File: src/CachePlugin.php
Comment (dbu) [inline src/CachePlugin.php]: afaik array_unique does not modify the input array. you need to re-assign the value to the field.

File: src/CachePlugin.php
Comment (dbu) [reply src/CachePlugin.php]: i agree with the approach. but i would directly use array_intersect here. as you assign it to a variable, its clear enough without the method. i would call the variable $nocacheDirectives.

File: src/CachePlugin.php
Comment (dbu) [reply src/CachePlugin.php:189]: agree with acrobat, this is handled below with the directives

File: src/CachePlugin.php
Comment (acrobat) [reply src/CachePlugin.php]: yes you're right! Fixed!

**Comment by acrobat** [comment]
@dbu Fixed your new comments!

File: src/CachePlugin.php
Comment (dbu) [inline src/CachePlugin.php]: this was indented on purpose, because its the list of fields in the $config array.

File: src/CachePlugin.php
Comment (dbu) [inline src/CachePlugin.php]: and this was left becaues it ends the list of options

File: src/CachePlugin.php
Comment (dbu) [inline src/CachePlugin.php]: i think this list should also have no-store

File: src/CachePlugin.php
Comment (dbu) [inline src/CachePlugin.php:339]: looks like adding no-cache is a change over current behaviour. i think we can consider that a bugfix. but please add a note to the changelog so people have a chance of seeing the change.

**Review by dbu — CHANGES_REQUESTED** [review]
we are almost there i think. no-cache was not handled previously, but i consider that a bugfix rather than a BC break.

File: src/CachePlugin.php
Comment (tuupola) [reply src/CachePlugin.php]: AFAIK you can actually cache the result if response has `no-cache` directive. It just means you must always revalidate before actually using the cached result. If origin server returns `304 Not Modified` you can use the cached result.

That said the easy implementation is just not to cache. Only difference is some bandwith might be wasted, but implementation is much easier.

https://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html#sec14.9.1

Comment (dbu) [reply src/CachePlugin.php]: right. lets open an issue to implement the full no-cache behaviour in the future. but rather than completely ignoring, completely not caching is at least correct according to the protocol.

Comment (acrobat) [reply src/CachePlugin.php]: Ok, I've added the no-store directive and left the no-cache for now

File: src/CachePlugin.php
Comment (acrobat) [reply src/CachePlugin.php]: Yeah I think this was fix for styleCi but indeed I've reverted the indent changes!

File: src/CachePlugin.php
Comment (acrobat) [reply src/CachePlugin.php]: See comment above

File: src/CachePlugin.php
Comment (acrobat) [reply src/CachePlugin.php:339]: I've added a line about this change in the changelog!

**Comment by acrobat** [comment]
I've did some changes according to the feedback and updated the changelog. @dbu & @Nyholm do you guys have time for a last review as I think we are almost ready here?

File: src/CachePlugin.php
Comment (dbu) [inline src/CachePlugin.php]: the second check also needs to have an isset. as is, this will be a fatal error if respect_response_cache_directives is not set. (and throw no exception if respect_cache_headers is set and the other is `[]`)

**Review by dbu — CHANGES_REQUESTED** [review]
thanks! almost there. i noticed one glitch in the constructor sanity check.

File: src/CachePlugin.php
Comment (acrobat) [reply src/CachePlugin.php]: good catch! Fixed!

**Comment by acrobat** [comment]
Fixed comment from @dbu and squashed/rebase the PR

**Review by dbu — APPROVED** [review]
looks good to me now. what do the other maintainers think?

File: src/CachePlugin.php
Comment (Nyholm) [reply src/CachePlugin.php:189]: Good

**Review by Nyholm — APPROVED** [review]
Nice. I really like this PR now =)

Thank you @acrobat

**Comment by dbu** [comment]
styleci is a bit weird. i guess we should remove that dot (i would be ok with it, but styleci should be green and i guess we can't easily make it allow just that one place)

**Comment by acrobat** [comment]
@dbu Yeah it was indeed a bit weird that's why I left it there. But removed the dot and now is styleci passing again!

**Comment by dbu** [comment]
thanks a lot! created https://github.com/php-http/documentation/issues/178 about the documentation. needs not a lot, but would be great to have it mentioned in the doc

**Comment by acrobat** [comment]
Yes indeed, that was still a todo for after the PR was merged. I will try to do that tonight!

File: src/CachePlugin.php
Comment (dbu) [reply src/CachePlugin.php]: created #27 for the full no-cache support

**Comment by acrobat** [comment]
Is there an ETA for the 1.3.0 release? As I would like to implement this feature in KnpLabs/php-github-api

**Comment by Nyholm** [comment]
There is no ETA yet. But I guess it will be sooner rather than later. 

Feel free to start a PR on KnpLabs/php-github-api and I'll be happy to review that one. I'll make sure both libraries release a new tag with this feature.

**Comment by acrobat** [comment]
Thanks for the info @Nyholm! I will start a PR to change the cachePlugin usage

**Comment by acrobat** [comment]
@sagikazarmark Do you maybe have some time to create a new release? Would be awesome to use this feature and the method caching feature from #24! Thanks!

**Comment by tuupola** [comment]
Wait a bit with a release. I just noticed that the regexp in #24 allowes `/` character which should not be there. I will make a bugfix PR tomorrow.

**Comment by acrobat** [comment]
ping @tuupola any time to fix this so we can have a release anytime soon?

**Comment by sagikazarmark** [comment]
> Do you maybe have some time to create a new release?

Sorry totally missed this one. My mailbox is quite full lately.

@tuupola ping me if you are done or need anything from me.

**Comment by tuupola** [comment]
Sorry for the delay. I will do it today.

**Pull request merged**
